nonthermal.cc

```diff
@@ -467,7 +467,64 @@ void init(const int my_rank, const int ndo_nonempty) {
     fflush(nonthermalfile);
   }
 
+  int LIST_LENGTH = MAX_NT_EXCITATIONS_STORED;
+#ifdef MPI_ON
+
+  MPI_Win nt_solution_window;
+
+  // We want to keep the nt_solution_struct in local but the nt_excitation_struct in shared memory
+  int rank, shm_rank;
+  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+
+  // Get the rank of the shared memory communicator
+  MPI_Comm shmcomm;
+  MPI_Comm_split_type(MPI_COMM_WORLD, MPI_COMM_TYPE_SHARED, 0, MPI_INFO_NULL, &shmcomm);
+
+  // Rank within the shared memory communicator
+  MPI_Comm_rank(shmcomm, &shm_rank);
+
+  // The max size of the vector is set to MAX_NT_EXCITATIONS_STORED hence the memory allocation is done for that size
+  int list_size = grid::get_npts_model() * sizeof(nt_excitation_struct) * LIST_LENGTH;
+
+  // Now add a printout telling us how much memory we are using / going to use
+  printout("[info] mem_usage: storing non-thermal frac_excitations_list for all allocated cells occupies %.3f MB\n",
+           list_size / 1024. / 1024.);
+
+  // MPI_Aint is the type of a variable able to contain a memory address.
+  MPI_Aint shared_mem_size = (shm_rank == 0) ? list_size : 0;
+
+  // Allocate shared memory for frac_excitations_list on local rank 0
+  nt_excitation_struct *shared_excitations_list = nullptr;
+  MPI_Win_allocate_shared(shared_mem_size, sizeof(nt_excitation_struct), MPI_INFO_NULL, shmcomm,
+                          &shared_excitations_list, &nt_solution_window);
+
+  // Obtain a pointer to the shared memory for non-local rank 0 processes
+  if (shm_rank != 0) {
+    int disp_unit;
+    MPI_Aint shared_size;
+    MPI_Win_shared_query(nt_solution_window, 0, &shared_size, &disp_unit, &shared_excitations_list);
+  }
+
+  nt_solution = static_cast<nt_solution_struct *>(calloc(grid::get_npts_model(), sizeof(nt_solution_struct)));
+
+  // Now we need to set the pointers for the nt_excitation_struct inside the nt_solution_struct to the shared memory
+  // Calculate how many excitations per nt_solution_struct
+  int excitations_per_solution = LIST_LENGTH;
+
+  for (int modelgridindex = 0; modelgridindex < grid::get_npts_model(); modelgridindex++) {
+    nt_excitation_struct *start = &shared_excitations_list[modelgridindex * excitations_per_solution];
+    nt_excitation_struct *end = &shared_excitations_list[((modelgridindex + 1) * excitations_per_solution) - 1];
+    nt_solution[modelgridindex].frac_excitations_list.assign(start, end);
```

Comment:
This is making a copy, so it looks like we are occupying the same memory per node as before (have you checked this in the JUWELS job report? did memory usage decrease?) plus the shared memory copy. Change the type of frac_excitations_list from std::vector to std::span so that it can hold pointers to the shared array instead of having its own storage with a copy of it.

Reply:
I’m most confused about this part: As I understand it, the current issue with the code arises during the frac_excitations_list resizing on the clusters during the nt_MPI_Bcast.  Each rank could theoretically require memory for MAX_NT_EXCITATIONS_STORED * the Total number of grid cells * the size of the vector struct. Hence, we are running out of memory.

Stuarts suggested fix: place the vector into shared memory, so that only one copy of each is needed per node, rather than per rank. Although this will require a very large structure to be allocated on the node, it should solve the memory issue. Hence after moving the structure into shared memory, we will need to assign pointers so that each rank knows the location of each cell in this large shared list. I don’t think the current implementation is doing this correctly then…

Another question on this section is does std::span when clear is applied does this remove the pointers to shared memory. I think that is the case when vectors are cleared.